Assets/MRTK/SDK/Experimental/Drawstring/Scripts/TwoHandleManipulator.cs

```diff
@@ -0,0 +1,937 @@
+            public HandleSide side;
+            private Vector3 initialGrabPointInPointer;
+
+            public PointerData(IMixedRealityPointer pointer, Vector3 worldGrabPoint, HandleSide side) : this()
+            {
+                this.pointer = pointer;
+                this.side = side;
+                this.initialGrabPointInPointer = Quaternion.Inverse(pointer.Rotation) * (worldGrabPoint - pointer.Position);
+            }
+
+            public bool IsNearPointer => pointer is IMixedRealityNearPointer;
+
+            /// Returns the grab point on the manipulated object in world space
+            public Vector3 GrabPoint => (pointer.Rotation * initialGrabPointInPointer) + pointer.Position;
+        }
+
+        private Dictionary<uint, PointerData> pointerIdToPointerMap = new Dictionary<uint, PointerData>();
+        private Quaternion objectToGripRotation;
+        private bool isNearManipulation;
+        private bool isManipulationStarted;
+
+        private Rigidbody rigidBody;
+        private bool wasKinematic = false;
+
+        private ConstraintManager constraints;
+
+        private bool IsOneHandedManipulationEnabled => manipulationType.HasFlag(ManipulationHandFlags.OneHanded) && pointerIdToPointerMap.Count == 1;
+        private bool IsTwoHandedManipulationEnabled => manipulationType.HasFlag(ManipulationHandFlags.TwoHanded) && pointerIdToPointerMap.Count > 1;
+
+        #endregion
+
+        #region MonoBehaviour Functions
+
+        private void Awake()
+        {
+            moveLogic = new ManipulationMoveLogic();
+            rotateLogic = new TwoHandRotateLogic();
+            elasticLogic = new ElasticHandleLogic();
+        }
+        private void Start()
+        {
+            rigidBody = HostTransform.GetComponent<Rigidbody>();
+            constraints = new ConstraintManager(gameObject);
+
+            leftLineOffset = lineDataProvider.transform.localPosition - leftHandle.localPosition;
+            rightLineOffset = (lineDataProvider.transform.localPosition + lineDataProvider.EndPoint.Position) - rightHandle.localPosition;
+            
+        }
+
+        private void Update()
+        {
+            if(pointerIdToPointerMap.Count == 0)
+            {
+                var elasticDistance = elasticLogic.Update(null, null, Time.time - lastUpdateTime);
+                lastUpdateTime = Time.time;
+                leftHandle.localPosition = Vector3.right * -elasticDistance / 2.0f;
+                rightHandle.localPosition = Vector3.right * elasticDistance / 2.0f;
+
+                UpdateLineData();
+            }
+        }
+
+        #endregion MonoBehaviour Functions
+
+        #region Private Methods
+        private Vector3 GetPointersGrabPoint()
+        {
+            Vector3 sum = Vector3.zero;
+            int count = 0;
+            foreach (var p in pointerIdToPointerMap.Values)
+            {
+                sum += p.GrabPoint;
+                count++;
```

Comment:
Nit: Do you need to maintain count? on line 412, can you just use the size()/count() of the map?